### ISSUE TITLE
ReadHeader() doesn't reset mapping delegates.

### ISSUE THREAD

**Issue by SGStino** [JoshClose/CsvHelper#1135]
We have a csv file that contains multiple sections.
```
[SectionA]
HeaderA, HeaderB, HeaderC
ValueA1, ValueB1, ValueC1
ValueA2, ValueB2, ValueC2
[SectionB]
HeaderB, HeaderA, HeaderC
ValueB3, ValueA3, ValueC3
ValueB4, ValueA4, ValueC4
```

i managed to split out the sections by looking at the Record in the parser context and calling ReadHeaders() again.

However, when reading the next records (GetRecord<MyType>()) the mapping still keeps the old indices, and as such, ValueB3 and ValueA3 are swapped around in MyType.

**Comment by JoshClose** [comment]
```cs
void Main()
{
	var s = new StringBuilder();
	s.AppendLine("[Section1]");
	s.AppendLine("IdA,NameA");
	s.AppendLine("1,one");
	s.AppendLine("[Section2]");
	s.AppendLine("NameB,IdB");
	s.AppendLine("two,2");
	using (var reader = new StringReader(s.ToString()))
	using (var csv = new CsvReader(reader))
	{
		csv.Configuration.ShouldSkipRecord = (record) => false;
		csv.Configuration.RegisterClassMap<TestMapA>();
		csv.Configuration.RegisterClassMap<TestMapB>();

		var aRecords = new List<A>();
		var bRecords = new List<B>();

		string section = string.Empty;
		var shouldReadHeader = false;
		while (csv.Read())
		{
			if (csv.Context.Record[0].StartsWith("[Section"))
			{
				section = csv.GetField(0);
				shouldReadHeader = true;
				continue;
			}
			
			if (shouldReadHeader)
			{
				csv.ReadHeader();
				shouldReadHeader = false;
				continue;
			}
			
			switch (section) 
			{
				case "[Section1]":
					aRecords.Add(csv.GetRecord<A>());
					break;
				case "[Section2]":
					bRecords.Add(csv.GetRecord<B>());
					break;
				default:
					throw new InvalidOperationException($"Section '{section}' is not valid.");
			}
		}

		aRecords.Dump();
		bRecords.Dump();
	}
}

public class A
{
	public int IdA { get; set; }
	public string NameA { get; set; }
}

public class B
{
	public int IdB { get; set; }
	public string NameB { get; set; }
}

public class TestMapA : ClassMap<A>
{
	public TestMapA()
	{
		Map(m => m.IdA);
		Map(m => m.NameA);
	}
}

public class TestMapB : ClassMap<B>
{
	public TestMapB()
	{
		Map(m => m.IdB );
		Map(m => m.NameB);
	}
}
```

![image](https://user-images.githubusercontent.com/114971/46750395-758db680-cc7d-11e8-9e7d-d0e60b8fe5a4.png)

**Comment by SGStino** [comment]
A and B are the same thing, just the order of the columns swapped. 

so it would more be:

```csharp
void Main()
{
	var s = new StringBuilder();
	s.AppendLine("[Section1]");
	s.AppendLine("x_Identifier,x_NameOne,x_NameTwo");
	s.AppendLine("1,one,two");
	s.AppendLine("[Section2]")
	s.AppendLine("x_Identifier,x_somethingElse,x_NameTwo,x_NameOne");
	s.AppendLine("2,3.14159,two,one");
	using (var reader = new StringReader(s.ToString()))
	using (var csv = new CsvReader(reader))
	{
		csv.Configuration.ShouldSkipRecord = (record) => false;
		csv.Configuration.RegisterClassMap<TestMapA>();
 		var allRecords = new List<A>();

		string section = string.Empty;
		var shouldReadHeader = false;
		while (csv.Read())
		{
			if (csv.Context.Record[0].StartsWith("[Section"))
			{
				section = csv.GetField(0);
				shouldReadHeader = true;
				continue;
			}
			
			if (shouldReadHeader)
			{
				csv.ReadHeader();
				shouldReadHeader = false;
				continue;
			}
			
			switch (section) 
			{
				case "[Section1]":
					allRecords.Add(csv.GetRecord<A>());
					break;
				case "[Section2]":
					allRecords.Add(csv.GetRecord<A>());
					break;
				default:
					throw new InvalidOperationException($"Section '{section}' is not valid.");
			}
		}

		allRecords.Dump();
 
	}
}

public class A
{
	public int Id { get; set; }
	public string NameOne { get; set; }
	public string NameTwo { get; set; }
}

public class TestMapA : ClassMap<A>
{
	public TestMapA()
	{
		Map(m => m.Id).Name("x_Identifier");
		Map(m => m.NameOne).Name("x_NameOne");		
                Map(m => m.NameTwo).Name("x_NameTwo");
	}
}

```

**Comment by JoshClose** [comment]
Oh I see. You're expecting `ReadHeader` to reset the indexes. I'll have to think about that.

For now you can create 2 mappings and register them both.

**Comment by SGStino** [comment]
the issue is that i don't know how many section there will be in advance, so i can't make a mapping per section.

I would expect that the mappings look at the current headers to look for the indexes, and ofc, if ReadHeader method sets the current headers, then i would expect the mappings to follow. 

And if the reader wasn't optimized for speed, i'd imagine it would work this way :)

**Comment by JoshClose** [comment]
How would `ReadHeader` know which mapping to use?

**Comment by SGStino** [comment]
How would the mapping know what index to use? by calling ReadHeader i guess?

so somewhere the mapping needs to be told to do that again?

**Comment by JoshClose** [comment]
I'm going to close this as I believe you have the recommended solution. If you have any more questions though, I'd be more than happy to answer them.

**Comment by SGStino** [comment]
if the mapper would use the current headers from the Context, it would be up-to-date with the new headers. But instead it caches it internaly in it's delegates. If there is any way to clear that cache i could do 
```
reader.ReadHeaders();
reader.MapperReset();
```
and the mapper would have to re-create it's delegates, form the headers that are present in the context at that point.

**Comment by JoshClose** [comment]
You can remove all registered class maps by doing this
```cs
reader.Configuration.UnregisterClassMap();
```

Creating the delegates is slow. Executing them is fast. This is why they are cached. You can register as many as you like

There is no way based on the header to tell which class map to use though. What if the headers are the same names, but in a different order? What if the headers are the same, but missing a column, or has an extra column? 

If you don't know how many headers there will be and how configurations, does that mean that you don't know the types either? Are you reading all fields as `string`?

**Comment by JoshClose** [comment]
Wait... I think I understand your issue here.

You're going to want to not use `GetRecord<T>()` and instead use `GetField<T>(string name)`. You don't want to have to build and compile a new delegate every time the header ordering changes.

There is an issue though. There is a `NamedIndex` dictionary that maps the header name to it's index/position in the CSV file. This will need to get reset before calling `ReadHeader()`.

You can do this
```cs
reader.Read();
if (this is a header row)
{
    reader.Context.NamedIndexes.Clear();
    reader.ReadHeader();
}
```

This should be called internally on every `ReadHeader()`, so I will fix that.

**Comment by JoshClose** [comment]
Sorry I wasn't understanding your issue clearly before.

This is fixed in version 9.2.2 on nuget. You don't have to clear the indexes anymore.
